### PR TITLE
Introducing blockchain proxies enum types

### PR DESCRIPTION
**Pull request by ii-cruz**
## What's in this pull request?
It introduces blockchain types by the creation of two enums - BlockchainProxy and BlockchainReadProxy. This is needed for the different sync strategies that require different blockchain implementations (eg. light and full sync).

## Pull request checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I have resolved any merge conflicts.

### CODE REVIEW

File: blockchain-proxy/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+tokio-stream = { version = "0.1", features = ["sync"] }
+futures = { package = "futures-util", version = "0.3" }
+
+nimiq-block = { path = "../primitives/block" }
+nimiq-blockchain = { path = "../blockchain" }
+nimiq-database = { path = "../database" }
+nimiq-genesis = { path = "../genesis" }
+nimiq-hash = { path = "../hash" }
+nimiq-primitives = { path = "../primitives" }
+nimiq-transaction = { path = "../primitives/transaction" }
```
Comment (jsdanielh) [inline blockchain-proxy/Cargo.toml:28]: Could you check all these dependencies? It looks to me like at least `nimiq-transaction` and `nimiq-database` are not used.

File: blockchain-proxy/src/blockchain_proxy.rs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+/// The `BlockchainReadProxy` implements `AbstractBlockchain` and allows to access common blockchain functions.
+/// It is a wrapper around read locked versions of our blockchain types.
+pub enum BlockchainReadProxy<'a> {
+    Full(Arc<RwLockReadGuard<'a, Blockchain>>),
+}
+
+impl<'a> AbstractBlockchain for BlockchainReadProxy<'a> {
+    fn network_id(&self) -> nimiq_genesis::NetworkId {
```
Comment (jsdanielh) [inline blockchain-proxy/src/blockchain_proxy.rs]: I'd suggest to do `use nimiq_genesis::NetworkId` in the beginning of the file and then only having `NetworkId` here. This also applies to the rest of the return types, like `nimiq_block::Block`, etc.

File: blockchain-proxy/src/blockchain_proxy.rs [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    ($self: ident, $t: ident, $f: ident $(, $arg:expr )*) => {
+        match $self {
+            $t::Full(ref blockchain) => AbstractBlockchain::$f(&***blockchain, $( $arg ),*),
+        }
+    };
+}
+
+/// The `BlockchainProxy` is our abstraction over multiple types of blockchains.
+/// Currently, it holds the:
+/// - (Full)Blockchain, which is capable of storing the full history, transactions, and full blocks.
```
Comment (jsdanielh) [inline blockchain-proxy/src/blockchain_proxy.rs:23]: I'd suggest to move this documentation to the member documentation (1 line below, just on top of `Full(...)`)

File: blockchain/src/blockchain/history_sync.rs [hunk 4]
```diff
@@ -376,14 +376,31 @@ impl Blockchain {
...
-        this.notifier
-            .notify(BlockchainEvent::HistoryAdopted(block_hash.clone()));
+        if let Err(e) = this
+            .notifier
+            .send(BlockchainEvent::HistoryAdopted(block_hash.clone()))
+        {
+            log::error!(
+                "Error sending the History Adopted event to the events notifier {}",
+                e
+            );
```
Comment (jsdanielh) [inline blockchain/src/blockchain/history_sync.rs:386]: I'd suggest to use the tracing style, which I believe would be:
```
log::error!(
    error = e,
    "Error sending the History Adopted event to the events notifier"
);
```
And the same for the log messages in this PR

File: consensus/Cargo.toml [hunk 5]
```diff
@@ -30,6 +30,7 @@ tokio-stream = { version = "0.1", features = ["sync"] }
 
 beserial = { path = "../beserial", features = ["derive"] }
 nimiq-block = { path = "../primitives/block" }
+nimiq-blockchain-proxy = { path = "../blockchain-proxy" }
 nimiq-bls = { path = "../bls" }
 nimiq-blockchain = { path = "../blockchain" }
```
Comment (jsdanielh) [inline consensus/Cargo.toml:35]: Is the `blockchain` still needed?

File: lib/src/client.rs [hunk 6]
```diff
@@ -1,12 +1,13 @@
 use std::sync::Arc;
 
+use nimiq_blockchain_proxy::BlockchainProxy;
 use nimiq_bls::cache::PublicKeyCache;
 use nimiq_nano_zkp::NanoZKP;
 use nimiq_primitives::policy::Policy;
 use parking_lot::{Mutex, RwLock};
 
 use nimiq_block::Block;
-use nimiq_blockchain::{AbstractBlockchain, Blockchain};
```
Comment (jsdanielh) [inline lib/src/client.rs:9]: I'd suggest to keep it like this and remove the added `use` in line 52.

File: rpc-server/src/dispatchers/consensus.rs [hunk 7]
```diff
@@ -1,10 +1,10 @@
 use std::sync::Arc;
 
 use async_trait::async_trait;
+use nimiq_blockchain::AbstractBlockchain;
```
Comment (jsdanielh) [inline rpc-server/src/dispatchers/consensus.rs]: It looks to me like it was ok as it was. I think the separation is for first import standard library components, then external and then our own.

**Review by jsdanielh — COMMENTED** [review]
Looks great. I only have some minor comments.

File: consensus/Cargo.toml (hunk 5 above)
Comment (ii-cruz) [reply consensus/Cargo.toml:35]: It is. The blockchain crate has the abstract blockchain trait, so we need to import it in several places in the consensus to interact with whatever is inside the blockchain proxy.
We are also directly using the blockchain type for at least for now. There are some match cases for specifying the behavior if it's a full blockchain when pushing a block.

File: blockchain-proxy/Cargo.toml (hunk 1 above)
Comment (ii-cruz) [reply blockchain-proxy/Cargo.toml:28]: They were all needed. I changed the implementation of Abstract Blockchain to make that clearer.

**Review by jsdanielh — APPROVED** [review]

**Comment by jsdanielh** [comment]
I'll rebase it to get it merged

**Pull request merged**